src/util/cow.h

```diff
@@ -0,0 +1,235 @@
+
+  copy_on_writet &operator=(const copy_on_writet &rhs)
+  {
+    auto copy(rhs);
+    swap(copy);
+    return *this;
+  }
+
+  copy_on_writet(copy_on_writet &&rhs)
+  {
+    swap(rhs);
+  }
+
+  copy_on_writet &operator=(copy_on_writet &&rhs)
+  {
+    swap(rhs);
+    return *this;
+  }
+
+  ~copy_on_writet() = default; // delete if the refcount is 1, like usual
+
+  void swap(copy_on_writet &rhs)
+  {
+    std::swap(t_, rhs.t_);
+  }
+
+  const T &read() const
+  {
+    return *t_;
+  }
+
+  T &write(bool mark_shareable)
+  {
+    if(pointee_use_count(*t_) != 1)
+    {
+      t_ = make_small_shared_ptr<T>(*t_);
+    }
+    INVARIANT(
+      pointee_use_count(*t_) == 1,
+      "mutable references to a COW pointer must be unique");
+    pointee_set_shareable(*t_, mark_shareable);
+    INVARIANT(
+      pointee_is_shareable(*t_) == mark_shareable,
+      "sharing flag must have expected value after obtaining mutable ref");
+    return *t_;
+  }
+
+  // Ideally these would be non-members, but they depend on private member t_
+
+  template <typename U>
+  bool operator==(const copy_on_writet<U> &rhs) const
+  {
+    return t_ == rhs.t_;
+  }
+
+  template <typename U>
+  bool operator!=(const copy_on_writet<U> &rhs) const
+  {
+    return t_ != rhs.t_;
+  }
+
+  template <typename U>
+  bool operator<(const copy_on_writet<U> &rhs) const
+  {
+    return t_ < rhs.t_;
+  }
+
+  template <typename U>
+  bool operator>(const copy_on_writet<U> &rhs) const
+  {
+    return t_ > rhs.t_;
+  }
+
+  template <typename U>
+  bool operator<=(const copy_on_writet<U> &rhs) const
+  {
+    return t_ <= rhs.t_;
+  }
+
+  template <typename U>
+  bool operator>=(const copy_on_writet<U> &rhs) const
+  {
+    return t_ >= rhs.t_;
+  }
+
+private:
+  small_shared_ptrt<T> t_;
+};
+
+////////////////////////////////////////////////////////////////////////////////
+
+template <typename Num>
+class copy_on_write_pointeet
+{
+public:
+  copy_on_write_pointeet() = default;
+
+  copy_on_write_pointeet(const copy_on_write_pointeet &)
+  {
+  }
+
+  copy_on_write_pointeet &operator=(const copy_on_write_pointeet &)
+  {
+    return *this;
+  }
+
+  copy_on_write_pointeet(copy_on_write_pointeet &&)
+  {
+  }
+
+  copy_on_write_pointeet &operator=(copy_on_write_pointeet &&)
+  {
+    return *this;
+  }
+
+  void increment_use_count()
+  {
+    INVARIANT(
+      is_shareable(),
+      "cannot increment the use count of a non-shareable reference");
+    use_count_ += 1;
+  }
+
+  void decrement_use_count()
+  {
+    INVARIANT(
+      is_shareable(),
+      "cannot decrement the use count of a non-shareable reference");
+    use_count_ -= 1;
+  }
+
+  Num use_count() const
+  {
+    return is_shareable() ? use_count_ : 1;
+  }
+
+  void set_shareable(bool u)
+  {
+    use_count_ = u ? 1 : unshareable;
+  }
+
+  bool is_shareable() const
+  {
+    return use_count_ != unshareable;
+  }
+
+protected:
+  ~copy_on_write_pointeet() = default;
+
+private:
+  static const Num unshareable;
```

Comment:
Given that I was very confused when reading this as I thought it would rely on default-initialisation to `0`: could you please add a comment that this will be set to `max` below?